presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java

```diff
@@ -1732,6 +1769,33 @@ else if (criteria instanceof JoinOn) {
             return output;
         }
 
+        private void verifyOnConditionReferencesRelatedFields(Relation rightRelation, Expression expression)
+        {
+            if (!(rightRelation instanceof AliasedRelation || rightRelation instanceof Table)) {
+                // TODO Add handling of TableSubquery, Lateral and Table for the case when columns in
+                //  JOIN ON condition specified without table name or alias (See #17382).
+                return;
+            }
+            if (!onConditionReferencesRelatedFields(expression, rightRelation)) {
+                String rightTableName;
+                if (rightRelation instanceof Table) {
+                    rightTableName = ((Table) rightRelation).getName().toString();
+                }
+                else {
+                    AliasedRelation aliasedRelation = (AliasedRelation) rightRelation;
+                    if (aliasedRelation.getRelation() instanceof Table) {
+                        rightTableName = ((Table) aliasedRelation.getRelation()).getName().toString();
+                    }
+                    else {
+                        rightTableName = aliasedRelation.getAlias().toString();
+                    }
+                }
+                String warningMessage = createWarningMessage(expression, format("JOIN ON condition(s) do not reference the joined table '%s' and other tables in the same " +
+                        "expression that can cause performance issues as it may lead to a cross join with filter", rightTableName));
```

Comment:
nit
```java
String warningMessage = createWarningMessage(
        expression,
        format(
                "JOIN ON condition(s) do not reference the joined table '%s' and other tables in the same " +
                        "expression that can cause performance issues as it may lead to a cross join with filter",
                rightTableName));